### ISSUE TITLE
site.posts.all() returns different post._ur'sl on two pages

### ISSUE THREAD

**Issue by thiagodebastos** [carrot/roots-dynamic-content#20]
## What
I am trying to apply a block, which includes a snippet for related posts, which simply lists other portfolio pieces for now.
The issue I am facing is that for some reason, Jade renders only one post in one page, but both in the other page. 

## How

``` jade
//- single-portfolio.jade
block content
	block creative-journey
	include related
```

``` jade
//- related.jade
each post in site.portfolio.all()
  li
    a(href= post._url)
      if post.relatedimageurl
        img(src= post.relatedimageurl)
      else
        img(src="http://lorempixel.com/325/250/abstract")
```
``` YAML
# dawnlister-therapy-centre.jade
# YAML front matter

date: 2014-01-15
title: Dawn Lister Therapy Centre.
description: Creative journey and case study for the Dawn Lister Therapy Centre website.
category: portfolio
tags: [website]
relatedimageurl: http://lorempixel.com/325/250/abstract/1
googlewebfonts: Lato|Open+Sans:400,700,800
_content: false

```

``` html
<!-- page1.html-->
<ul class="work-grid">
    <li><a href="/portfolio/dawn-lister-therapy-centre.html"><img src="http://lorempixel.com/325/250/abstract/1"></a></li>
</ul>

```

``` html
<!-- page2.html-->
<ul class="work-grid">
    <li><a href="/portfolio/dawn-lister-therapy-centre.html"><img src="http://lorempixel.com/325/250/abstract/1"></a></li>
    <li><a href="/portfolio/solid-nutrition-london.html"><img src="http://lorempixel.com/325/250/abstract/2"></a></li>
</ul>

```

## Why
Once this is working properly, I want to be able to list 3 related posts at the bottom of my page. I had this working [[1]](http://thiagodebastos.com/work/solid-nutrition-london/) [[2]](http://thiagodebastos.com/work/dawn-lister-therapy-centre/) in Jekyll and am in the process of converting everything into Roots because it rocks! 

### TLDR; What I Noticed
Rendered posts are not always in the same order. If I save a file while it is being watched, the contents of printed posts will swap around. I have confirmed this behaviour to exist in the output of content.json. I am not sure if this is related but I thought I would add the observation. If it is a bug, I am more than happy to move the issue where it belongs, if not here.

#### Extra Context
``` coffee
// app.coffee - irrelevant info such as locals have been ommited
module.exports =
  ignores: ['readme.md', '_templates/**/*', '**/_*', '**/*.sublime*',  '_includes/**/*', '.gitignore', '.editorconfig', 'ship.*conf']

  dump_dirs: [''] // fits my project structure

  extensions: [
    dynamic_content(write: 'content.json'),
    js_pipeline(files: ['assets/js/*.coffee', 'assets/js/*.js']),
    css_pipeline(files: 'assets/css/*.styl'),
    yaml()
  ]

  jade:
    pretty: true
    basedir: '/Users/Thiago/Code/thiagodebastos-roots/'
```

**Comment by jenius** [comment]
Hey so here's the background on this and why it doesn't work as expected. When rendering dynamic content, roots will run through all the content as quickly as possible. This means that you can't expect a certain order to persist every time, and also that from inside any given piece of dynamic content, you cannot expect all the other pieces of dynamic content to be available. The only way to do this would be to double compile everything, which would be a waste. This means that you cannot access all other dynamic content directly from within any given piece of dynamic content, just from the layouts they are rendered into.

Because we have had this ask before specifically for "related posts", you can [export all the dynamic content as json](https://github.com/carrot/roots-dynamic-content/blob/master/docs/intro.md#exporting-dynamic-content-as-json) then load that up with javascript on a per-page basis.

**Comment by thiagodebastos** [comment]
Thank you for clearing this up! I have learned quite a bit about JS and JSON thanks to all of this hacking and slashing. I appreciate your clean and concise replies. Once my site is done, the first blog post I will make will have to be a guide for other noobs to tame Roots.  😷

**Comment by jenius** [comment]
Would be great, we're still working on guides and tutorial videos if you are interesting in contributing we'd be happy to host some of it on roots' site!

**Comment by thiagodebastos** [comment]
That sounds great, I'd be glad to help. I'll get back to you in the next week or so :)

**Comment by toabi** [comment]
I kind of abused the exporting to have *all* dynamic content on the generating side. I the extension write it out as a json file, and then in the `app.config` I use fs.readFileSync to read it. Caveat: On the first run the file doesn't exist yet, so it needs always two runs when changing stuff in those files, but that's okay.

**Comment by jenius** [comment]
Huh, interesting. That's working fine and pulling in all the content?

**Comment by toabi** [comment]
Sure, I mean the generated `public/blog.json` contains all the dynamic content I want.

It's a bit hacky and the *issue* is, that it needs two runs until it's up to date when blog content was changed. But that's okay, because I can stay all "server-side" in my jade templates without bothering building a client-side JS-Navigation.

My code at the top of `application.json`: 

    try
        blogDataJson = JSON.parse(fs.readFileSync('public/blog.json'))
        blogItems = _.sortByOrder(blogDataJson.items, 'date', false)
    catch
        console.log 'blog.json not generated yet, please re-run roots!'
        blogItems = []

And then just `module.exports.locals.blogItems = blogItems`.

**Comment by jenius** [comment]
Oh this is interesting for sure. Yeah it seems like a number of people are having this issue, the only way I can think of that will solve it natively would involve a double compile, as this does, as this is the nature of the issue. I'm a little too restricted on time at this point in my life to tackle this myself, but it seems like we should work on a slightly cleaner solution, even if it does take longer to compile as you are not the only one who needs this.
